simpleviewinc/tap-events-force
# Zen 371 booking button states

**Ticket**: [ZEN-371](https://jira.simpleviewtools.com/browse/ZEN-371)

**NOTES**
* A new key called `bookingStopped` has been added to the testData in some places.
* This is used to represent a session that has had it's booking stopped, but booking was allowed at some point
* Currently Events Force is not sending us any way to determine this state. I have reached out to them, but have not heard back
    * This will be updated once I get a response from them on what the key should be

## Context
* The grid container and grid items have buttons related to booking sessions
* The buttons should reflect the current state based on the session and attendees
* Specs doc: https://docs.google.com/document/d/1bZIMvNhg6fsU5Tf1edLcoBfpu8rR9qQLGf0Kx-DGMrw/edit#

## Goal
* Display the correct booking state for each session based on the session and attendees

## Updates
* `configs/eslintrc.config.js`
    * Add `c` for the `color` rule shortcut in re-theme
* `src/actions/session/selectSession.js`
    * Updated to accept a custom attendees array
    * This allows passing an empty array, which should remove attendees from the session
* `src/assets/icons/evf/elements/bookingCheck.js`
    * Added booking checkmark icon
* `src/assets/icons/evf/elements/digit.js`
    * This is not technically an icon, but it's used like one, and acts like one; so it made sense to go here 
* `src/assets/icons/evf/index.js`
    * Added the Icon name as a property of the wrapper component so it can be accessed outside the wrapper
* `src/components/button/bookingButton.js`
    * Updated to display the look of the booking button based on the booking state
* `src/components/button/evfButton.js`
    * Added disabled and children prop
    * Allows overwriting the text prop with child components
    * Allows setting the button to be disabled from outside the button
* `src/components/testData/testData.js`
    * Added a select component to switch the test data to different states
    * This makes it easier / quicker to test the state of the app based on different test data
    * Right now I've only added differnt booking states, but other states could be added
* `src/constants/values.js`
    * Updated the session booking state `available` to be `select`
    * This matches the text on the booking button like the other states
* `src/hooks/booking/useBookingState.js`
    * Custom hook to build the booking model for the session
* `src/hooks/booking/useBookingTimeConflicts.js`
    * Custom hook to get any time conflicts for attendees and the current session
* `src/mocks/eventsforce/bookingStates/*`
    * Added testData for different booking states
* `src/models/session/bookingState.js`
    * Model for a sessions booking state
* `src/theme/components/button/evfButton.js`
    * Added styles for the different booking states
* `src/utils/models/sessions/bookingStateFactory.js`
    * Builds booking state models based on teh passed in props
    * Creates helper methods for each booking state, that returns a booking state model
* `src/utils/models/sessions/getBookingState.js`
    * Added check for `bookingStopped`
    * This will most likely need to be updated once Events Force responds with what it should be



## Testing
**Notes**
* Pull the docker package  => `keg evf pack run docker.pkg.github.com/simpleviewinc/keg-packages/tap:zen-371-booking-button-states`
* Navigate to [http://local.kegdev.xyz/](http://local.kegdev.xyz/)
* Open the Test Data (JSON) editor
    * On the left you should see a select dropdown
    * This will allow you to switch to different booking states with pre-defined test data
    * [Looks like this](http://snpy.in/mnr2E3)
* Each option follows a naming convection to help test a particular state
    * The pattern goes like this => `[booking state]-[booking-mode]-[reason]`
        * `booking-state` - This is the state the button should be in when this option is selected
        * `booking-mode` - This is the booking mode based on the number of attendees
            * single (1 attendee)
            * grouped ( more then 1 attendee )
        * `reason` - This is just a simple reason for why it's in this state based on the test data
            * Really I just needed a `key` in the object to define the state and this is what I came up with
            * It really just gives a hint of what the test data represents
                * For example
                    * `conflict` => An attendee has a time conflict between 2 sessions
                    * `override` => There are attendees on the booking and waiting list, so the `on waiting list` state overrides the `selected` state
    * Remember to hit the apply button after switching the booking state from the select box

**Test 1**
* Go through each one of the different states from the select dropdown
* Validate the data in the test-data editor with what is being showing on screen for each session
    * Selecting different states should reflect a different state on the Booking Button
    * The look of the booking button should match the [mock-ups](https://www.figma.com/file/kTLTvkxNF7ziNe9dQee88n/sessions_v2?node-id=0%3A1)

**Test 2**
* Ensure
    * When selecting a group booking state, the group booking modal still opens
    * When selecting a disabled booking state, the booking button is disabled
    * When selecting the `(on waiting list | selected) and single mode`, and clicking the booking button
        * It should call the selectSession action, with an empty array
        * [This should print out in the inspector](http://snpy.in/54mmg5)

**Test 3**
* Make modifications to the test data, to try and get different states for the sessions
    * See if you can find a variation I might have missed
